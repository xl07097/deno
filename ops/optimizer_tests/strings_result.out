#[allow(non_camel_case_types)]
///Auto-generated by `deno_ops`, i.e: `#[op]`
///
///Use `op_string_length::decl()` to get an op-declaration
///you can include in a `deno_core::Extension`.
pub struct op_string_length;
#[doc(hidden)]
impl op_string_length {
    pub const fn name() -> &'static str {
        stringify!(op_string_length)
    }
    #[allow(clippy::not_unsafe_ptr_arg_deref)]
    pub extern "C" fn v8_fn_ptr<'scope>(
        info: *const deno_core::v8::FunctionCallbackInfo,
    ) {
        let info = unsafe { &*info };
        let scope = &mut unsafe { deno_core::v8::CallbackScope::new(info) };
        let args = deno_core::v8::FunctionCallbackArguments::from_function_callback_info(
            info,
        );
        let rv = deno_core::v8::ReturnValue::from_function_callback_info(info);
        Self::v8_func(scope, args, rv);
    }
    pub const fn decl<'scope>() -> deno_core::OpDecl {
        deno_core::OpDecl {
            name: Self::name(),
            v8_fn_ptr: Self::v8_fn_ptr as _,
            enabled: true,
            fast_fn: None,
            is_async: false,
            is_unstable: false,
            is_v8: false,
            force_registration: false,
            arg_count: 1usize as u8,
        }
    }
    #[inline]
    #[allow(clippy::too_many_arguments)]
    fn call(string: &str) -> Result<u32, AnyError> {
        Ok(string.len() as u32)
    }
    pub fn v8_func<'scope>(
        scope: &mut deno_core::v8::HandleScope<'scope>,
        args: deno_core::v8::FunctionCallbackArguments,
        mut rv: deno_core::v8::ReturnValue,
    ) {
        let ctx = unsafe {
            &*(deno_core::v8::Local::<deno_core::v8::External>::cast(args.data()).value()
                as *const deno_core::_ops::OpCtx)
        };
        let arg_0 = match deno_core::v8::Local::<
            deno_core::v8::String,
        >::try_from(args.get(0usize as i32)) {
            Ok(v8_string) => deno_core::serde_v8::to_utf8(v8_string, scope),
            Err(_) => {
                return deno_core::_ops::throw_type_error(
                    scope,
                    format!("Expected string at position {}", 0usize),
                );
            }
        };
        let arg_0 = arg_0.as_ref();
        let result = Self::call(arg_0);
        let op_state = ::std::cell::RefCell::borrow(&*ctx.state);
        op_state.tracker.track_sync(ctx.id);
        match result {
            Ok(result) => {
                rv.set_uint32(result as u32);
            }
            Err(err) => {
                let exception = deno_core::error::to_v8_error(
                    scope,
                    op_state.get_error_class_fn,
                    &err,
                );
                scope.throw_exception(exception);
            }
        };
    }
}
