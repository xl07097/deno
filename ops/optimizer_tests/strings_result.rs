// https://github.com/denoland/deno/issues/16979
fn op_string_length(string: &str) -> Result<u32, AnyError> {
  Ok(string.len() as u32)
}
