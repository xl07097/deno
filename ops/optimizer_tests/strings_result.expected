=== Optimizer Dump ===
returns_result: true
has_ref_opstate: false
has_rc_opstate: false
has_fast_callback_option: false
needs_fast_callback_option: true
fast_result: Some(U32)
fast_parameters: [V8Value, SeqOneByteString]
transforms: {0: Transform { kind: SeqOneByteString(Ref), index: 0 }}
is_async: false
fast_compatible: false
